src/Extensions/Compatability/Validation/Code.cs

```diff
@@ -0,0 +1,489 @@
+		/// <param name="argumentName">The local name of the argument.</param>
+		/// <param name="tagId">Tag identifier to log; leave null if no logging is required.</param>
+		/// <returns>Returns the argument value as is if validation succeeds, otherwise an exception is thrown.</returns>
+		/// <exception cref="ArgumentNullException">Thrown if the supplied argument <paramref name="argumentValue"/> is null.</exception>
+		public static T ExpectsObject<T>([ValidatedNotNull] T? argumentValue, string argumentName, EventId? tagId) where T : class
+		{
+			return ExpectsArgumentNotNull(argumentValue, argumentName, tagId);
+		}
+
+
+		/// <summary>
+		/// Validate object's state
+		/// </summary>
+		/// <param name="state">Object's state</param>
+		/// <param name="errorMessage">Error message</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed</param>
+		/// <returns>True if state evaluates to true; false otherwise</returns>
+		public static bool Validate(bool state, string errorMessage, EventId? tagId)
+		{
+			if (!state)
+			{
+				LogError(tagId, errorMessage);
+			}
+
+			return state;
+		}
+
+
+		/// <summary>
+		/// Validate object's state
+		/// </summary>
+		/// <param name="state">Object's state</param>
+		/// <param name="errorMessage">Error message</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed</param>
+		/// <param name="parameters">logging parameters</param>
+		/// <returns>True if state evaluates to true; false otherwise</returns>
+		public static bool Validate(bool state, EventId? tagId, string errorMessage, params object[] parameters)
+		{
+			if (!state)
+			{
+				LogError(tagId, errorMessage, parameters);
+			}
+
+			return state;
+		}
+
+
+		/// <summary>
+		/// Checks that the collection is not null, is not empty, and does not contain nulls
+		/// </summary>
+		/// <typeparam name="T">The type of the collection</typeparam>
+		/// <param name="argumentValue">The argument value.</param>
+		/// <param name="argumentName">Name of the argument.</param>
+		/// <param name="tagId">Tag Id to log, leave null if no logging is needed</param>
+		/// <returns>True if the argument <paramref name="argumentValue"/> is not null, is not empty, and does not contain nulls; false otherwise.</returns>
+		public static bool ValidateNotEmptyAndAllNotNull<T>([NotNullWhen(true)]ICollection<T>? argumentValue, string argumentName, EventId? tagId = null)
+			where T : class
+		{
+			if (!ValidateArgument(argumentValue, argumentName, tagId))
+			{
+				return false;
+			}
+
+			if (!argumentValue.Any() || argumentValue.Any(x => x == null))
+			{
+				LogError(tagId, ValidationFailed, AllErrorMessage, argumentName);
+				return false;
+			}
+
+			return true;
+		}
+
+
+		/// <summary>
+		/// Checks that the enumerable argument is not null and doesn't contain any nulls
+		/// </summary>
+		/// <remarks>Be careful to not pass enumerables that can be enumerated only once</remarks>
```

Comment:
>Be careful to not pass enumerables that can be enumerated only once [](start = 15, length = 67)

For all of these `Validate` and `Expect` methods taking `IEnumerable` objects, it would be worth calling out in the `<remarks>` that these entail iterating over all of the objects and consequently can have a significant performance overhead.

Reply:
I think in future those methods need to be changed to accept different types (ex. ICollection) or deleted completely